systems/controllers/osc/osc_tracking_data.h

```diff
@@ -0,0 +1,428 @@
+  Eigen::VectorXd ddy_des_;
+
+  // `state_` is the finite state machine state when the tracking is enabled
+  // If `state_` is empty, then the tracking is always on.
+  std::vector<int> state_;
+
+ private:
+  // Check if we should do tracking in the current state
+  void UpdateTrackingFalg(int finite_state_machine_state,
+                  double time_since_last_state_switch);
+
+  // Updaters of feedback output, jacobian and dJ/dt * v
+  virtual void UpdateYAndError(const Eigen::VectorXd& x_w_spr,
+                               KinematicsCache<double>& cache_w_spr,
+                               const RigidBodyTree<double>& tree_w_spr) = 0;
+  virtual void UpdateYdot(const Eigen::VectorXd& x_w_spr,
+                          KinematicsCache<double>& cache_w_spr,
+                          const RigidBodyTree<double>& tree_w_spr) = 0;
+  virtual void UpdateJ(const Eigen::VectorXd& x_wo_spr,
+                       KinematicsCache<double>& cache_wo_spr,
+                       const RigidBodyTree<double>& tree_wo_spr) = 0;
+  virtual void UpdateJdotV(const Eigen::VectorXd& x_wo_spr,
+                           KinematicsCache<double>& cache_wo_spr,
+                           const RigidBodyTree<double>& tree_wo_spr) = 0;
+  // Finalize and ensure that users construct OscTrackingData derived class
+  // correctly.
+  virtual void CheckDerivedOscTrackingData() = 0;
+
+  // Trajectory name
+  std::string name_;
+
+  // Dimension of the traj
+  int n_r_;
+
+  // PD control gains
+  Eigen::MatrixXd K_p_;
+  Eigen::MatrixXd K_d_;
+
+  // Cost weights
+  Eigen::MatrixXd W_;
+
+  // Trajectory info
+  bool traj_is_const_;
+  bool traj_has_exp_;
+
+  // Store the fixed position if the traj is a const
+  Eigen::VectorXd fixed_position_;
+
+  // A period when we don't apply control
+  // (starting at the time when fsm switches to a new state)
+  double period_of_no_control_ = 0;  // Unit: seconds
+
+  // cache
+  bool track_at_current_step_;
+  int state_idx_ = 0;
+  Eigen::VectorXd ddy_command_;
+};
+
+
+// TaskSpaceTrackingData is still a virtual class
+class TaskSpaceTrackingData : public OscTrackingData {
+ public:
+  TaskSpaceTrackingData(std::string name, int n_r,
+                        Eigen::MatrixXd K_p,
+                        Eigen::MatrixXd K_d,
+                        Eigen::MatrixXd W,
+                        bool traj_is_const,
+                        bool traj_has_exp);
+
+  TaskSpaceTrackingData() {}  // Default constructor
+
+ protected:
+  // `body_index_w_spr` is the index of the body
+  // `body_index_wo_spr` is the index of the body
+  // If `body_index_w_spr_` is empty, `body_index_wo_spr_` replaces it.
+  std::vector<int> body_index_w_spr_;
+  std::vector<int> body_index_wo_spr_;
+
+ private:
+};
+
+
+/// TransTaskSpaceTrackingData is used when we want to track a trajectory
+/// (translational position) in the task space.
+
+/// `track_center_of_mass_` should be set to true when we track center of mass.
```

Comment:
I think it's probably better to have it within `TransTaskSpaceTrackingData`, because conceptually center of mass position lives in translational task space. It might be easier for the user to understand the whole `TrackingData` family if it stay inside `TransTaskSpaceTrackingData`? Additionally, `ComTrackingData`'d share methods `AddPointToTrack`